### PR TITLE
WFP-1466

### PR DESCRIPTION
**Pull request by carlov20**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt [hunk 1]
```diff
@@ -1,49 +1,39 @@
...
+import org.springframework.web.reactive.function.client.awaitBodyOrNull
+import org.springframework.web.reactive.function.client.bodyToFlow
 import reactor.core.publisher.Mono
 import uk.gov.justice.digital.hmpps.hmppsallocations.domain.RiskPredictor
 import uk.gov.justice.digital.hmpps.hmppsallocations.domain.RoshSummary
 
 class AssessRisksNeedsApiClient(private val webClient: WebClient) {
 
-  fun getRosh(crn: String): Mono<RoshSummary> {
+  suspend fun getRosh(crn: String, authToken: String): RoshSummary? {
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt:15]: couldn't figure out how use filter on the actual webclient when initialising so had to pass through from controller

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/WebClientUserEnhancementConfiguration.kt [hunk 2]
```diff
@@ -77,40 +68,45 @@ class WebClientUserEnhancementConfiguration(
...
+    val service: ReactiveOAuth2AuthorizedClientService = InMemoryReactiveOAuth2AuthorizedClientService(clients)
+    val manager = AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager(clients, service)
 
-    val defaultClientCredentialsTokenResponseClient = DefaultClientCredentialsTokenResponseClient()
+    val reactiveClientCredentialsTokenResponseClient = WebClientReactiveClientCredentialsTokenResponseClient()
 
-    val authentication = SecurityContextHolder.getContext().authentication
+    reactiveClientCredentialsTokenResponseClient.addParametersConverter { grantRequest ->
+      val parameters: MultiValueMap<String, String> = LinkedMultiValueMap()
+      runBlocking {
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/WebClientUserEnhancementConfiguration.kt:79]: this does not work, username is always null here

File: src/main/resources/application.yml [hunk 3]
```diff
@@ -86,4 +86,9 @@ management:
...
 
 oauth:
-  endpoint.url: http://localhost:8090/auth
\ No newline at end of file
+  endpoint.url: http://localhost:8090/auth
+
+azure:
+  application-insights:
+    web:
+      enabled: false
```
Comment (carlov20) [inline src/main/resources/application.yml:94]: needed to stop servlet being initialised

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/AuthHeaderWebFilter.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+import org.springframework.http.HttpHeaders
+import org.springframework.stereotype.Component
+import org.springframework.web.server.ServerWebExchange
+import org.springframework.web.server.WebFilter
+import org.springframework.web.server.WebFilterChain
+import reactor.core.publisher.Mono
+
+@Component
+class AuthHeaderWebFilter : WebFilter {
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/AuthHeaderWebFilter.kt:11]: to allow the auth token to be passed through to ARN API, need to set it in the context here to reference later

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/AuthTokenFilterFunction.kt [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+class AuthTokenResponseMono(
+  private val request: ClientRequest,
+  private val next: ExchangeFunction
+) : Mono<ClientResponse>() {
+
+  override fun subscribe(subscriber: CoreSubscriber<in ClientResponse>) {
+    val context = subscriber.currentContext()
+    val requestBuilder = ClientRequest.from(request)
+    requestBuilder.header(HttpHeaders.AUTHORIZATION, context.get(HttpHeaders.AUTHORIZATION))
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/AuthTokenFilterFunction.kt:26]: referenced later after setting in the context in the `AuthHeaderWebFilter`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/HmppsAllocationsExceptionHandler.kt [hunk 6]
```diff
@@ -9,11 +9,11 @@ import org.springframework.http.HttpStatus.METHOD_NOT_ALLOWED
 import org.springframework.http.ResponseEntity
 import org.springframework.http.converter.HttpMessageConversionException
 import org.springframework.security.access.AccessDeniedException
-import org.springframework.web.HttpRequestMethodNotSupportedException
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/HmppsAllocationsExceptionHandler.kt:12]: this is a Servlet exception, no more servlets in reactive. the equivalent is `MethodNotAllowedException`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/WebClientUserEnhancementConfiguration.kt [hunk 7]
```diff
@@ -32,15 +31,9 @@ class WebClientUserEnhancementConfiguration(
...
-  fun assessRisksNeedsWebClientUserEnhancedAppScope(builder: WebClient.Builder, httpServletRequest: HttpServletRequest): WebClient {
+  fun assessRisksNeedsWebClientUserEnhancedAppScope(builder: WebClient.Builder): WebClient {
     return builder.baseUrl(assessRisksNeedsApiRootUri)
-      .filter { request: ClientRequest, next: ExchangeFunction ->
-        val filtered = ClientRequest.from(request)
-          .header(HttpHeaders.AUTHORIZATION, httpServletRequest.getHeader(HttpHeaders.AUTHORIZATION))
-          .build()
-        next.exchange(filtered)
-      }
+      .filter(AuthTokenFilterFunction())
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/WebClientUserEnhancementConfiguration.kt:36]: uses the pass through auth filter to set it in the request to send to ARN API

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/WebClientUserEnhancementConfiguration.kt [hunk 8]
```diff
@@ -64,53 +56,58 @@ class WebClientUserEnhancementConfiguration(
...
-  private fun authorizedClientManagerUserEnhanced(clients: ClientRegistrationRepository?): OAuth2AuthorizedClientManager {
-    val service: OAuth2AuthorizedClientService = InMemoryOAuth2AuthorizedClientService(clients)
-    val manager = AuthorizedClientServiceOAuth2AuthorizedClientManager(clients, service)
+  private fun authorizedClientManagerUserEnhanced(clients: ReactiveClientRegistrationRepository?, builder: WebClient.Builder): ReactiveOAuth2AuthorizedClientManager {
+    val service: ReactiveOAuth2AuthorizedClientService = InMemoryReactiveOAuth2AuthorizedClientService(clients)
+    val manager = AuthorizedClientServiceReactiveOAuth2AuthorizedClientManager(clients, service)
+    val reactiveClientCredentialsTokenResponseClient = WebClientReactiveClientCredentialsTokenResponseClient()
 
-    val defaultClientCredentialsTokenResponseClient = DefaultClientCredentialsTokenResponseClient()
+    reactiveClientCredentialsTokenResponseClient.setWebClient(builder.filter(userEnhancedTokenRequestProcessor()).build())
```
Comment (carlov20) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/config/WebClientUserEnhancementConfiguration.kt:76]: reactive equivalent of `setRequestEntityConverter` where you can modify the request before requesting a client creds oauth token

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt [hunk 9]
```diff
@@ -1,49 +1,35 @@
...
-          is MissingRiskError -> Mono.empty()
-          else -> Mono.error(ex)
-        }
-      }
+        { Mono.empty() }
+      ).awaitBodyOrNull()
   }
 
-  fun getRiskPredictors(crn: String): Flux<RiskPredictor> {
+  fun getRiskPredictors(crn: String): Flow<RiskPredictor> {
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt:25]: Doesn't this need a `suspend`?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt [hunk 10]
```diff
@@ -1,49 +1,35 @@
...
-        { Mono.error(MissingRiskError("No risk summary found for $crn")) }
-      )
-      .bodyToMono(RoshSummary::class.java)
-      .onErrorResume { ex ->
-        when (ex) {
-          is MissingRiskError -> Mono.empty()
-          else -> Mono.error(ex)
-        }
-      }
+        { Mono.empty() }
```
Comment (sp-watson) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt:21]: I'm assuming there are tests for this - the docs do not make it obvious what the behaviour is when combining Mono  and `awaitBodyOrNull`

File: src/main/resources/application.yml (hunk 3 above)
Comment (sp-watson) [reply src/main/resources/application.yml:94]: But doesn't this stop `requests` being sent to App Insights - that's what happened in the lookup service

**Review by sp-watson — COMMENTED** [review]
I am assuming all the config changes match those in other fully reactive projects.

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt (hunk 9 above)
Comment (carlov20) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt:25]: `bodyToFlow()` isn't a suspend function so the compiler said suspend was redundant

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt (hunk 10 above)
Comment (carlov20) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/client/AssessRisksNeedsApiClient.kt:21]: https://github.com/ministryofjustice/hmpps-allocations/blob/main/src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsallocations/integration/unallocatedcases/GetCaseRisksByCrnTest.kt#L95

File: src/main/resources/application.yml (hunk 3 above)
Comment (sp-watson) [reply src/main/resources/application.yml:94]: Update: Integration tests won't run without this. We suspect it's because we haven't gone to Spring Boot 3

Comment (carlov20) [reply src/main/resources/application.yml]: going to open telemetry seems to have done the trick https://github.com/ministryofjustice/hmpps-allocations/pull/332/commits/2e5edc529719631f850d0d798f73a4a7f1265039, would be good to see what is logged in app insights after this goes in

**Review by sp-watson — APPROVED** [review]

**Pull request merged**